KAFKA-4081: Consumer API consumer new interface commitSync does not v…

mimaison opened the pull request:
…erify the validity of offset

Commit throws InvalidOffsetException if the offset is negative

hachikuji commented:
@mimaison Adding a new error code would unfortunately require a KIP. Can we do a simpler client-side check for now?

mimaison commented:
@hachikuji By doing the check on the client side, if one of the offsets to commit is invalid we'll have to throw an error and cancel the whole request. Is that acceptable ?

Should we do attempt to do a KIP to properly fix this on the server side ?

hachikuji commented:
@mimaison I'm not sure I see the complication. It seems preferable to catch the invalid offsets before the request is sent to the coordinator and raise an `IllegalArgumentException`, though it would be more ideal if the server also had a check. A KIP may be the way to go if we think the lack of a server-side check is a significant problem. What did you think about my previous suggestion to not allow commits on offsets which are greater than the current position?

mimaison committed:
KAFKA-4081: Consumer API consumer new interface commitSync does not verify the validity of offset

Commit will now fail with an IllegalArgumentException when trying to commit an offset that is either negative or larger than the consumer's current position

Developed with @edoardocomar

mimaison commented:
Thanks for the feedback. I've updated the PR and moved the offset validation on the client side. I've also added a check for offsets larger than the current position.

mimaison commented:
The Jenkins build timed out, I don't think it was caused by our changes.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1231,19 +1244,32 @@ public void seekToEnd(Collection<TopicPartition> partitions) {
         }
     }
 
+    private void positions(Set<TopicPartition> partitions) {
```
It seems a little weird to have no return value for `positions()` while `position()` returns a long. Would it be better to name it something like `updatePositions()`?

mimaison replied:
Good point, fixed

becketqin commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1154,12 +1155,24 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
+        for (Map.Entry<TopicPartition, OffsetAndMetadata> entry : offsets.entrySet()) {
+            OffsetAndMetadata offsetAndMetadata = entry.getValue();
+            Long position = subscriptions.position(entry.getKey());
+            if (offsetAndMetadata.offset() < 0 || offsetAndMetadata.offset() > position) {
```
We probably should document this behavior. I am not sure if a larger than position offset is necessarily invalid. I saw sometimes users would want to just reset their offset of a group to a certain position which might be higher than the current position. Arguably this can also be achieved by calling `seek()` then `commit()`, but it maybe a little inconvenient if there are a lot of partitions.

mimaison replied:
By doing seek() and commit() it makes it explicit that we want to commit a larger offset and the inconvenience is pretty small.

mimaison commented:
@becketqin @hachikuji Do you have any more comments ?

hachikuji commented:
@mimaison Thanks for pinging me. I'll take another look today.

hachikuji commented:
@mimaison Thanks for the patch. I think raising an exception when the offset is negative makes total sense, but I wonder if we should be a bit more conservative about the other case since it is a backwards incompatible change. Also, one of the unfortunate side effects of calling `updatePositions` is that `commitAsync` can now block. I'm probably a little inclined to leave the behavior as it is or do a short KIP to get some wider feedback. What do you think?

mimaison commented:
I agree, all the complexity comes from the checks for offsets larger than the current position and as you said it also slightly changes the behaviour.

I'll update the PR to only keep the negative offset check

mimaison commented:
@hachikuji  I've updated the PR as suggested, can you have another look ?
Thanks !

hachikuji commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -1230,6 +1230,27 @@ class PlaintextConsumerTest extends BaseConsumerTest {
   }
 
+  @Test
+  def testCommitInvalidOffsets() {
```
This might be a little cheaper to test in `KafkaConsumerTest` if you can stand the mocking. If possible, it would be nice to try to avoid adding unnecessarily to the already painful build times.

becketqin commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -419,7 +419,7 @@ class PlaintextConsumerTest extends BaseConsumerTest {
       consumer.commitAsync(Map(tp -> new OffsetAndMetadata(i)).asJava, callback)
 
-    awaitCommitCallback(consumer, callback, count=count)
+    awaitCommitCallback(consumer, callback, successCount=count)
```
nit: "successCount=count" => "successCount = count".

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -584,6 +584,9 @@ public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception
         for (Map.Entry<TopicPartition, OffsetAndMetadata> entry : offsets.entrySet()) {
             OffsetAndMetadata offsetAndMetadata = entry.getValue();
+            if (offsetAndMetadata.offset() < 0) {
+                return RequestFuture.failure(new IllegalArgumentException("Invalid offset: " + offsetAndMetadata.offset()));
```
Do you think it makes sense to try and catch this earlier (like in `KafkaConsumer`), or did you want to ensure that the autocommit path is covered as well?

mimaison replied:
Having the check here allows to only check in one place and also avoids looping another time.

becketqin commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -1230,6 +1230,27 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+
+      // async commit - negative offset
+      this.consumers.head.commitAsync(negativeOffset, callback)
+      awaitCommitCallback(this.consumers.head, callback, timeoutMs=5000)
```
It seems that the default `successCount` is 1 here so we are always going to wait for 5 seconds?

hachikuji reviewed (commented):
Thanks for the patch. Left a comment and question.

mimaison commented:
@hachikuji We've moved the tests into ConsumerCoordinatorTest. We tried to mock things in KafkaConsumerTest but ran into a bunch of issues.

hachikuji commented:
LGTM. Thanks for the patch. @becketqin Do the updates look good to you?

becketqin commented:
Thanks for the patch. LGTM.

hachikuji commented:
@mimaison I changed the title when merging. Hopefully it's reasonable.
